modules/internal/ChapelReduce.chpl

```diff
@@ -52,7 +52,16 @@ module ChapelReduce {
   
   proc chpl__sumType(type eltType) type {
     var x: eltType;
-    return (x + x).type;
+    if isArray(x) {
+      type xET = x.eltType;
+      type xST = chpl__sumType(xET);
+      if xET == xST then
+        return eltType;
+      else
+        return [x.domain] xST;
+    } else {
+      return (x + x).type;
+    }
```

Comment:
Good questions.

First, if (x+x).type worked, I wouldn't have bothered with this change. The reason it does not work is that (x+x) produces an _ir_ type, or anyway something similar. That does not work with the way the types of reduce-intent-related temps are inferred.

Why just for + ? That's because there is such a function only for +. The other operations are handled via _prod_id, _land_id, et al. (in modules/standard/Types.chpl). I suspect that those work for a smaller set of types than chpl__sumType().